common/cmake/devices.cmake

```diff
@@ -929,6 +929,164 @@ function(ADD_FPGA_TARGET_BOARDS)
   endforeach()
 endfunction()
 
+function(ADD_BITSTREAM_TARGET)
+  set(options)
+  set(oneValueArgs USE_FASM)
+  set(multiValueArgs INCLUDED_TARGETS)
+  cmake_parse_arguments(
+    ADD_BITSTREAM_TARGET
+    "${options}"
+    "${oneValueArgs}"
+    "${multiValueArgs}"
+    ${ARGN}
+  )
+
+
+  set(NAME ${ADD_BITSTREAM_TARGET_NAME})
+  set(INCLUDED_TARGETS ${ADD_BITSTREAM_TARGET_INCLUDED_TARGETS})
+  set(USE_FASM ${ADD_BITSTREAM_TARGET_USE_FASM})
+
+  if("${USE_FASM}" STREQUAL "")
+    set(USE_FASM TRUE)
+  endif()
+  # Generate bitstream
+  # -------------------------------------------------------------------------
+  set(ALL_OUT_FASM "")
+  if(${USE_FASM})
+    foreach(TARGET ${INCLUDED_TARGETS})
+      get_target_property_required(BOARD ${TARGET} BOARD)
+      get_target_property_required(DEVICE ${BOARD} DEVICE)
+      get_target_property_required(DEVICE_TYPE ${DEVICE} DEVICE_TYPE)
+      get_target_property(USE_OVERLAY ${DEVICE_TYPE} USE_OVERLAY)
+      get_target_property_required(OUT_FASM ${TARGET} OUT_FASM)
+      if ("${FASM_TO_BIT_DEPS}" STREQUAL "FASM_TO_BIT_DEPS-NOTFOUND")
+        set(FASM_TO_BIT_DEPS "")
+      endif()
+
+      string(APPEND ALL_OUT_FASM ${OUT_FASM} )
+
+      if(USE_OVERLAY)
+        if (NOT OVERLAY)
+          set(OVERLAY ${TARGET})
+        else()
+          message(FATAL_ERROR "More than one overlay device for ${BOARD}")
+        endif()
+      endif()
+    endforeach()
+
+    if (NOT OVERLAY)
+      list(LENGTH ${INCLUDED_TARGETS} TARGETS_LENGTH)
+      if(NOT ${TARGETS_LENGTH} EQUAL 0)
+        message(FATAL_ERROR "Multiple devices but no overlay for ${BOARD}")
+      endif()
+      set(OVERLAY ${INCLUDED_TARGETS})
+    endif()
+
+    get_target_property_required(OVERLAY_BOARD ${OVERLAY} BOARD)
+    get_target_property_required(OVERLAY_DEVICE ${OVERLAY_BOARD} DEVICE)
+    get_target_property_required(PACKAGE ${OVERLAY_BOARD} PACKAGE)
+    get_target_property_required(PYTHON3 env PYTHON3)
+
+    get_target_property(FASM_TO_BIT_EXTRA_ARGS ${OVERLAY_BOARD} FASM_TO_BIT_EXTRA_ARGS)
+    if ("${FASM_TO_BIT_EXTRA_ARGS}" STREQUAL "FASM_TO_BIT_EXTRA_ARGS-NOTFOUND")
+      set(FASM_TO_BIT_EXTRA_ARGS "")
+    endif()
+
+    get_target_property_required(OUT_FASM ${OVERLAY} OUT_FASM)
+    get_target_property_required(OUT_FASM_MERGED ${OVERLAY} OUT_FASM_MERGED)
+    get_target_property_required(OUT_LOCAL ${OVERLAY} OUT_LOCAL)
```

Comment:
Can you explain why you need OUT_LOCAL from the OVERLAY?  Seems unneed?

Reply:
My thought is to generate the merged bitstream in the directory of the overlay target, so I was getting OUT_LOCAL from OVERLAY to know where to write the bitstream.  Is there a better way to do this?

Reply:
Supply an argument rather than adding properties?

Reply:
Also, I don't think it makes any sense to output the merged bitstream in the overlay directory.  Why use that output folder?

Reply:
I don't think there's a specific reason to use that output folder, it's just what I knew how to do.  Still pretty new to cmake.  What's the correct way to add a new output folder for the merged fasm and bitstream generation?

Reply:
You have a couple options.  The most flexible would be to take an argument for the output path.  When using the function for the add_fpga_target, you could specific old location.  For new joint 1+N bitstreams, you would want to create a folder with the output target name, so something like `${CMAKE_CURRENT_BINARY_DIR}/${NAME}`